mlrun/common/secrets.py

```diff
@@ -0,0 +1,147 @@
+# Copyright 2023 Iguazio
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from abc import ABC, abstractmethod
+
+import mlrun.common.schemas
+
+
+class SecretProviderInterface(ABC):
+    @abstractmethod
+    def store_auth_secret(
+        self, username: str, access_key: str, namespace=""
+    ) -> (str, mlrun.common.schemas.SecretEventActions):
+        pass
+
+    @abstractmethod
+    def delete_auth_secret(self, secret_ref: str, namespace=""):
+        pass
+
+    @abstractmethod
+    def read_auth_secret(self, secret_name, namespace="", raise_on_not_found=False):
+        pass
+
+    @abstractmethod
+    def store_project_secrets(
+        self, project, secrets, namespace=""
+    ) -> (str, mlrun.common.schemas.SecretEventActions):
+        pass
+
+    @abstractmethod
+    def delete_project_secrets(self, project, secrets, namespace=""):
+        pass
+
+    @abstractmethod
+    def get_project_secret_keys(self, project, namespace="", filter_internal=False):
+        pass
+
+    @abstractmethod
+    def get_project_secret_data(self, project, secret_keys=None, namespace=""):
+        pass
+
+    @abstractmethod
+    def get_secret_data(self, secret_name, namespace=""):
+        pass
+
+
+class InMemorySecretProvider(SecretProviderInterface):
+    def __init__(self):
+        # project -> secret_key -> secret_value
+        self.project_secrets_map = {}
+        # ref -> secret_key -> secret_value
+        self.auth_secrets_map = {}
+        # secret-name -> secret_key -> secret_value
+        self.secrets_map = {}
+
+    @staticmethod
+    def get_auth_secret_name(username: str, access_key: str) -> str:
+        return f"secret-ref-{username}-{access_key}"
+
+    def store_auth_secret(
+        self, username: str, access_key: str, namespace=""
+    ) -> (str, mlrun.common.schemas.SecretEventActions):
+        secret_ref = self.get_auth_secret_name(username, access_key)
+        self.auth_secrets_map.setdefault(secret_ref, {}).update(
+            self._generate_auth_secret_data(username, access_key)
+        )
+        return secret_ref, mlrun.common.schemas.SecretEventActions.created
+
+    @staticmethod
+    def _generate_auth_secret_data(username: str, access_key: str):
```

Comment:
Move after interface public functions. Making class more readable
Also, why bothering with static decoration if is private? I think makes it more complicated than needed
Especially when it will eventually require the self

Reply:
Will move to end. I don't see static decoration as complicated - this function has class-related logic that doesn't depend on a specific instance, so it makes sense for it to be static. There's no functionality gained by it though one may call it directly without having to hold a class instance, but I see documentation value in it.